services/proxy/pkg/userroles/oidcroles.go

```diff
@@ -0,0 +1,227 @@
+			err := errors.New("invalid role in claims")
+			logger.Error().Err(err).Interface("claimValue", cri).Msg("Is not a valid string.")
+			return nil, err
+		}
+		id, ok := claimValueToRoleID[cr]
+		if !ok {
+			logger.Debug().Str("role", cr).Msg("No mapping for claim role. Skipped.")
+			continue
+		}
+		roleIDsFromClaim = append(roleIDsFromClaim, id)
+	}
+	logger.Debug().Interface("roleIDs", roleIDsFromClaim).Msg("Mapped claim roles to roleids")
+
+	switch len(roleIDsFromClaim) {
+	default:
+		err := errors.New("too many roles found in claims")
+		logger.Error().Err(err).Msg("Only one role per user is allowed.")
+		return nil, err
+	case 0:
+		err := errors.New("no role in claim, maps to a ocis role")
+		logger.Error().Err(err).Msg("")
+		return nil, err
+	case 1:
+		// exactly one mapping. This is right
+		break
+	}
+	assignedRoles, err := loadRolesIDs(ctx, user.GetId().GetOpaqueId(), ra.roleService)
+	if err != nil {
+		logger.Error().Err(err).Msg("Could not load roles")
+		return nil, err
+	}
+	if len(assignedRoles) > 1 {
+		err := errors.New("too many roles assigned")
+		logger.Error().Err(err).Msg("The user has too many roles assigned")
+		return nil, err
+	}
+	logger.Debug().Interface("assignedRoleIds", assignedRoles).Msg("Currently assigned roles")
+	if len(assignedRoles) == 0 || (assignedRoles[0] != roleIDsFromClaim[0]) {
+		logger.Debug().Interface("assignedRoleIds", assignedRoles).Interface("newRoleIds", roleIDsFromClaim).Msg("Updating role assignment for user")
+		newctx, err := ra.prepareAdminContext()
+		if err != nil {
+			logger.Error().Err(err).Msg("Error creating admin context")
+			return nil, err
+		}
+		if _, err = ra.roleService.AssignRoleToUser(newctx, &settingssvc.AssignRoleToUserRequest{
+			AccountUuid: user.GetId().GetOpaqueId(),
+			RoleId:      roleIDsFromClaim[0],
+		}); err != nil {
+			logger.Error().Err(err).Msg("Role assignment failed")
+			return nil, err
+		}
+	}
+	enc, err := encodeRoleIDs(roleIDsFromClaim)
+	if err != nil {
+		logger.Error().Err(err).Msg("Could not encode loaded roles")
+		return nil, err
+	}
+
+	if user.Opaque == nil {
+		user.Opaque = &types.Opaque{
+			Map: map[string]*types.OpaqueEntry{
+				"roles": enc,
+			},
+		}
+	} else {
+		user.Opaque.Map["roles"] = enc
+	}
+
+	return user, nil
+}
+
+// ApplyUserRole it looks up the user's role in the settings service and adds it
+// user's opaque data
+func (ra oidcRoleAssigner) ApplyUserRole(ctx context.Context, user *cs3.User) (*cs3.User, error) {
+	roleIDs, err := loadRolesIDs(ctx, user.Id.OpaqueId, ra.roleService)
+	if err != nil {
+		ra.logger.Error().Err(err).Msgf("Could not load roles")
+		return nil, err
+	}
+
+	enc, err := encodeRoleIDs(roleIDs)
+	if err != nil {
+		ra.logger.Error().Err(err).Msg("Could not encode loaded roles")
+		return nil, err
+	}
+
+	if user.Opaque == nil {
+		user.Opaque = &types.Opaque{
+			Map: map[string]*types.OpaqueEntry{
+				"roles": enc,
+			},
+		}
+	} else {
+		user.Opaque.Map["roles"] = enc
+	}
```

Comment:
This could be replaced with `utils.AppendJSONToOpaque` either. Nice little helper function, isn't it?

Reply:
Oh. Yeah. And I always forget about it. Thanks (once more) for the hint. :smile: